exchanges/coinbasepro/coinbasepro_websocket.go

```diff
@@ -46,423 +83,464 @@ func (c *CoinbasePro) WsConnect() error {
+		return "", errCantDecodePrivKey
+	}
+	key, err := x509.ParseECPrivateKey(block.Bytes)
+	if err != nil {
+		return "", err
+	}
+	nonce, err := common.GenerateRandomString(16, "1234567890ABCDEF")
+	if err != nil {
+		return "", err
+	}
+	head := map[string]any{"kid": creds.ClientID, "typ": "JWT", "alg": "ES256", "nonce": nonce}
+	headJSON, err := json.Marshal(head)
+	if err != nil {
+		return "", err
+	}
+	headEncode := base64URLEncode(headJSON)
+	body := map[string]any{"iss": "cdp", "nbf": time.Now().Unix(), "exp": time.Now().Add(time.Minute * 2).Unix(), "sub": creds.ClientID, "aud": "retail_rest_api_proxy"}
+	if uri != "" {
+		body["uri"] = uri
+	} else {
+		c.jwtLastRegen = time.Now()
 	}
-	err := c.Websocket.Conn.SendJSONMessage(context.TODO(), request.Unset, r)
-	if err == nil {
-		err = c.Websocket.RemoveSubscriptions(c.Websocket.Conn, subs...)
+	bodyJSON, err := json.Marshal(body)
+	if err != nil {
+		return "", err
+	}
+	bodyEncode := base64URLEncode(bodyJSON)
+	hash := sha256.Sum256([]byte(headEncode + "." + bodyEncode))
+	sig, err := ecdsa.SignASN1(rand.Reader, key, hash[:])
+	if err != nil {
+		return "", err
+	}
+	sigEncode := base64URLEncode(sig)
+	return headEncode + "." + bodyEncode + "." + sigEncode, nil
+}
+
+// getTimestamp is a helper function which pulls a RFC3339-formatted timestamp from a byte slice
+// of JSON data
+func getTimestamp(rawData []byte) (time.Time, error) {
+	data, _, _, err := jsonparser.Get(rawData, "timestamp")
+	if err != nil {
+		return time.Time{}, err
+	}
+	timestamp, err := time.Parse(time.RFC3339, string(data))
+	if err != nil {
+		return time.Time{}, err
+	}
+	return timestamp, nil
+}
+
+// processBidAskArray is a helper function that turns WebsocketOrderbookDataHolder into arrays
+// of bids and asks
+func processBidAskArray(data *WebsocketOrderbookDataHolder) (bids, asks orderbook.Tranches, err error) {
+	bids = make(orderbook.Tranches, 0, len(data.Changes))
+	asks = make(orderbook.Tranches, 0, len(data.Changes))
+	for i := range data.Changes {
+		change := orderbook.Tranche{Price: data.Changes[i].PriceLevel, Amount: data.Changes[i].NewQuantity}
+		switch data.Changes[i].Side {
+		case "bid":
+			bids = append(bids, change)
+		case "offer":
+			asks = append(asks, change)
+		default:
+			return nil, nil, fmt.Errorf("%w %v", order.ErrSideIsInvalid, data.Changes[i].Side)
+		}
+	}
+	bids.SortBids()
+	asks.SortAsks()
+	return bids, asks, nil
+}
+
+// statusToStandardStatus is a helper function that converts a Coinbase Pro status string to a
+// standardised order.Status type
+func statusToStandardStatus(stat string) (order.Status, error) {
+	switch stat {
+	case "received":
+		return order.New, nil
+	case "open":
+		return order.Active, nil
+	case "done":
+		return order.Filled, nil
+	case "match":
+		return order.PartiallyFilled, nil
+	case "change", "activate":
+		return order.Active, nil
+	default:
+		return order.UnknownStatus, fmt.Errorf("%w %v", errUnrecognisedStatusType, stat)
+	}
+}
```

Comment:
These may have changed since you did this:
https://docs.cdp.coinbase.com/advanced-trade/docs/ws-channels#user-channel
I note that CB seems to be doing things in upper case and I'm not seeing case handling anywhere